### PR TITLE
Spec the ingress ports addition.

### PR DESCRIPTION
**Pull request by oponder**
To make this story a bit more concrete. RFC @marians @xh3b4sd

Already working on the `kubernetesd` part, but wanted to pause a bit and see if we are aligned for the final result of this, in terms of what will actually be added / changed in the API.

https://github.com/giantswarm/api-spec/issues/50

### CODE REVIEW

**Comment by oponder** [comment]
I'm not sure about the `port_mappings` field name by the way. I chose that because that is kind of how I understand this concept. If there is a more appropriate term, please lets use that.

here is what a `GET /v4/clusters/{id}/` could look like now:

```
              {
                "id": "wqtlq",
                "create_date": "2017-03-03T10:50:45.949270905Z",
                "api_endpoint": "https://api.wqtlq.example.com",
                "name": "Just a Standard Cluster",
                "release_version": "2.5.16",
                "kubernetes_version": "",
                "owner": "acme",
                "workers": [
                  {
                    "memory": {"size_gb": 2},
                    "storage": {"size_gb": 20},
                    "cpu": {"cores": 4},
                    "labels": {
                      "beta.kubernetes.io/arch": "amd64",
                      "beta.kubernetes.io/os": "linux",
                      "ip": "10.3.11.2",
                      "kubernetes.io/hostname": "worker-1.x882ofna.k8s.gigantic.io",
                      "nodetype": "hicpu"
                    }
                  },
                  {
                    "memory": {"size_gb": 8},
                    "storage": {"size_gb": 20},
                    "cpu": {"cores": 2},
                    "labels": {
                      "beta.kubernetes.io/arch": "amd64",
                      "beta.kubernetes.io/os": "linux",
                      "ip": "10.3.62.2",
                      "kubernetes.io/hostname": "worker-2.x882ofna.k8s.gigantic.io",
                      "nodetype": "hiram"
                    }
                  }
                ],
                "port_mappings": [
                  {
                    "ingressPort": 30010,
                    "lbPort": 30020,
                    "protocol": "http"
                  },
                  {
                    "ingressPort": 30011,
                    "lbPort": 30021,
                    "protocol": "https"
                  },
                ]
              }
```

**Comment by oponder** [comment]
> The mapping is from the ingress controller on the host cluster to the service port of the guest cluster pods (which run on the host cluster as well).

> ok why then is it called lbPort?

> Because the `lbPort` is what customers have to configure in their LB on prem.
> Then it goes [LB] -> [host cluster ingress] -> [guest cluster service].

Still doesn't really compute to me :(
It doesn't really line up with the way we've been talking about it in issues so far.

I want to explain this concept well though. 

Is there anyone else that can take a stab at explaining what is actually going on here to me, and what names would make sense to people looking at this? @puja108 perhaps?

**Comment by oponder** [comment]
Like, do my descriptions make sense in definitions.yaml?

File: definitions.yaml [hunk 1]
```diff
@@ -136,7 +136,29 @@ definitions:
...
+        description: |
+          Reveals the ports on the load balancer that are mapped to this cluster's ingress
+          and which protocol that port supports. Only shown and relevant on our on-prem KVM clusters.
+        items:
+          $ref: '#/definitions/V4PortMappingDefinition'
 
+  V4PortMappingDefinition:
+    type: object
+    properties:
+      ingress_port:
```
Comment (xh3b4sd) [inline definitions.yaml]: I am not sure this should be exposed. This information is more or less useless because the customer cannot do anything with it AFAIK because this is for the k8s service manifest of the guest cluster which runs in our host cluster. The ports are also always the same. They will always be `30010` for HTTP and `30011` for HTTPS IIRC. See https://github.com/giantswarm/kubernetesd/blob/a790b6be50f18dfe1dd6f7fbd504e8ff2c333fc0/service/creator/ingress.go#L60-L103. 

The actual ingress-controller ports are configured somewhere else AFAIK.

File: definitions.yaml [hunk 2]
```diff
@@ -136,7 +136,29 @@ definitions:
...
+        description:
+          The port on this cluster's ingress.
+        type: integer
+      load_balancer_port:
+        description: |
+          The port on the load balancer that is mapped to this cluster's ingress.
```
Comment (xh3b4sd) [inline definitions.yaml]: It maybe helps to use the concepts of customer/company relationship and the concept of our host cluster wording. I am biased but something like this is most clear to me personally. 
 
```
The port on the customer's load balancer that is mapped to the Giant Swarm host cluster ingress-controller.
```

Comment (oponder) [reply definitions.yaml]: hmm thx, it gets a bit wordy for me, and doesn't help me understand.

The host cluster shouldn't even have anything to do with anything here. We're looking at details of a guest cluster. 

So this is the port on the load balancer that lets you talk to the ingress of your guest cluster. whether or not the host cluster is in between there is not relevant right?

Comment (xh3b4sd) [reply definitions.yaml]: > The host cluster shouldn't even have anything to do with anything here.

This is how it works. I just state facts and we should be as clear as possible to make customers understand what they are doing. IMO transparency is a good thing. I also only tried to help. You can gladly do what you want here. Maybe lets just go with the best we come up and iterate on it. When we notice there is a common misunderstanding over and over again we can adapt. 

> So this is the port on the load balancer that lets you talk to the ingress of your guest cluster. whether or not the host cluster is in between there is not relevant right?

It is relevant to me in order to understand how it works. When you want to hide this to the customer, fine with me. Your sentence I quoted here is quite accurate in summary I guess.

Comment (oponder) [reply definitions.yaml]: Thanks for the help! Wasn't rejecting it out right, just trying to get closer to the real understanding, but from the outsiders perspective.

Comment (puja108) [reply definitions.yaml]: I think there's a misunderstanding there. The host cluster has quite a lot to do with this as we are talking about onprem here.

In onprem the host cluster ingress controller forwards TCP ports to the guest cluster ingress controller. The guest cluster ingress controller needs 2 of those (http/https). Thus, needs to forward traffic on 2 ports from his (hardware) lb to host cluster IC.

The thing we call lbport is actually a port on the host cluster (behind which the host cluster ingress controller sits). So it is not the port of the customer's load balancer. On the customer's load balancer that is usually sth like IP:443 and IP:80 (where IP is the dedicated IP for a single guest cluster).

So a typical example for a guest cluster https request would look like following:

optional outside DNS -> https://ingress.clusterid.basedomain (DNS)-> IP:443 (on hw lb) -> 30021 (on host cluster) -> host cluster IC TCP service for cluster x -> 30011 (on guest cluster ) -> guest cluster IC https -> (service (only for endpoint lookup in IC -> ) pod -> container

File: definitions.yaml (hunk 1 above)
Comment (puja108) [reply definitions.yaml]: true, this is internal and always the same, also the customer can look this up inside their cluster, as this is basically their IC config

File: definitions.yaml [hunk 3]
```diff
@@ -136,7 +136,29 @@ definitions:
...
+          and which protocol that port supports. Only shown and relevant on our on-prem KVM clusters.
+        items:
+          $ref: '#/definitions/V4PortMappingDefinition'
 
+  V4PortMappingDefinition:
+    type: object
+    properties:
+      ingress_port:
+        description:
+          The port on this cluster's ingress.
```
Comment (puja108) [inline definitions.yaml]: external port of the cluster's ingress controller

File: definitions.yaml (hunk 2 above)
Comment (puja108) [reply definitions.yaml]: Maybe `This is the port on the host cluster ingress controller that needs to be set up to receive traffic for this cluster.`

Comment (puja108) [reply definitions.yaml]: or even minus the `to be set up` it needs to receive traffic, how you solve that, is up to you, and you most probably know if you are looking for this.

**Comment by oponder** [comment]
How about "The port on the host cluster that will forward traffic to the guest cluster" ?

**Comment by oponder** [comment]
Anyone still opposed to anything in here? Can I get an approve?

**Comment by oponder** [comment]
@marians @puja108 @xh3b4sd This spec is finally implemented. Though while implementing I ended up scoping the `port_mappings` under a `kvm` object since this is a `kvm` specific piece of info.
